### PR TITLE
Support recording breadcrumbs

### PR DESCRIPTION
**Pull request by Cawllec**
Adds breadcrumbs to ruby, allowing users to log data to be added to subsequent notification payloads, as well as keeping an audit trail of previous notifications.

### CODE REVIEW

File: lib/bugsnag/breadcrumbs/breadcrumb.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+            attr_accessor :name
+            attr_accessor :type
+            attr_accessor :timestamp
+            attr_accessor :metadata
+
+            def initialize(name, type, metadata = {})
+                self.timestamp = Time.now.utc.strftime "%Y-%m-%dT%H:%MZ"
+
+                if !name || name.length === 0
```
[the remark this answers is not in the document]
Comment (keeganlow) [reply lib/bugsnag/breadcrumbs/breadcrumb.rb]: we should most likely use ruby's == for this sort of integer equality check. if we expected name.length to potentially return a range/class/regex, === would make more sense.

**Comment by Cawllec** [comment]
I'm updated the breadcrumb and reporter objects to have more logical structures and thread-safety.  I haven't sorted the indentation or auto-attachment yet

File: lib/bugsnag/breadcrumbs/recorder.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+module Bugsnag
+    module Breadcrumbs
+
+        class Recorder
```
[the remark this answers is not in the document]
Comment (snmaynard) [reply lib/bugsnag/breadcrumbs/recorder.rb]: The thread safety you have added isn't correct. You want to use thread locals like in requestData. We can chat about it Monday if you like - let me know. We effectively need a unique list of breadcrumbs for each thread and to clear them at the end of each request

[the remark this answers is not in the document]
Comment (Cawllec) [reply lib/bugsnag/breadcrumbs/recorder.rb]: Ah, I get now what you meant.  I'll have a look at it tomorrow and catch up with you at some-point to discuss it

**Comment by Cawllec** [comment]
The remaining Rubocop issues on this repo are to do with `Report` class length, and `notify` method length/branch conditions.  As a task for refactoring is in the works i'd suggest adding exceptions to the issues for now and removing them through the refactor.  Thoughts @kattrali?

**Comment by bdewater** [comment]
Hey 👋  I came across breadcrumbs in the docs and would love to see this PR bring support to the Ruby gem. Is this still planned to be merged?

**Comment by Cawllec** [comment]
Closing this for now as work is being done on this in a different way.

**Pull request closed without merging**